### PR TITLE
[dune] [stdlib] Build the standard library natively with Dune.

### PR DESCRIPTION
**Pull request by ejgallego**
This completes a pure Dune bootstrap of Coq, last commit is the relevant one.

This requires Dune >= 2.5.0

### CODE REVIEW

File: theories/dune [hunk 1]
```diff
@@ -0,0 +1,35 @@
+(coq.theory
+ (name Coq)
+ (public_name coq.Coq)
+ ; (package coq)
+ (synopsis "Coq's Standard Library")
+ (flags -q -allow-sprop)
```
Comment (Zimmi48) [inline theories/dune]: The `-allow-sprop` flag is not needed anymore.

Comment (ejgallego) [reply theories/dune]: Good point, thanks.

**Comment by ejgallego** [comment]
This is (finally!) ready modulo a new dune release that includes https://github.com/ocaml/dune/pull/2053

File: theories/dune [hunk 2]
```diff
@@ -0,0 +1,38 @@
+(coq.theory
+ (name Coq)
+ (package coq)
+ (synopsis "Coq's Standard Library")
+ (flags -q)
+ ; (mode native)
+ (boot)
+ ; (per_file
+ ;  (Init/*.v -> -boot))
```
Comment (Zimmi48) [inline theories/dune]: Why keeping commented out lines?

File: theories/dune [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+   coq.plugins.nsatz
+   coq.plugins.omega
+
+   coq.plugins.zify
+   coq.plugins.micromega
+
+   coq.plugins.funind
+
+   coq.plugins.ssreflect
+   coq.plugins.derive))
```
Comment (Zimmi48) [inline theories/dune]: I am not sure I understand the reason behind the order and the spacing.

**Review by Zimmi48 — APPROVED** [review]
This is :rocket: 

Too bad that this will have to wait for Dune 2.5, hopefully soon enough though!
My comments are only on the form. Otherwise, this LGTM.

File: theories/dune (hunk 2 above)
Comment (ejgallego) [reply theories/dune]: Mainly as a reminder to update this once dune has per_file flags, I'll add a comment.

File: theories/dune (hunk 3 above)
Comment (ejgallego) [reply theories/dune]: Oh, actually the order corresponds [IIANM] to the order the plugins are needed in the stdlib, from first to last. I dunno why I did it that way.

File: Makefile.dune [hunk 4]
```diff
@@ -1,122 +1,109 @@
...
-COQTOP_FILES=ide/idetop.bc ide/coqide_main.bc checker/coqchk.bc
-PLUGIN_FILES=$(wildcard plugins/*/*.mlpack)
-PRINTER_FILES=dev/top_printers.cma
-QUICKBYTE_TARGETS=$(COQTOP_FILES) $(PLUGIN_FILES:.mlpack=.cma) $(PRINTER_FILES) topbin/coqtop_byte_bin.bc
-QUICKOPT_TARGETS=$(COQTOP_FILES:.bc=.exe) $(PLUGIN_FILES:.mlpack=.cmxs) $(PRINTER_FILES:.cma=.cmxa) topbin/coqtop_bin.exe
-
-quickbyte: voboot
-	dune build $(DUNEOPT) $(QUICKBYTE_TARGETS)
-
-quickopt: voboot
```
Comment (Zimmi48) [inline Makefile.dune:86]: I frequently use `make -f Makefile.dune quickopt` to make sure that I haven't broken any ML code. What should I use instead? `dune build @check`?

File: Makefile.dune [hunk 5]
```diff
@@ -1,122 +1,109 @@
...
 
-fmt: voboot
-	dune build @fmt --auto-promote
+# We define this target as to override Make's built-in one
+install:
+	@echo "To install Coq using dune, use 'dune install -p PACKAGE' where"
+	@echo "PACKAGE is any of the packages defined by opam files in the root dira"
+
+fmt:
+	dune build @fmt
```
Comment (Zimmi48) [inline Makefile.dune:100]: You recently added the `--auto-promote`. Any reason to backtrack now?

**Review by Zimmi48 — COMMENTED** [review]
Will need a bump in nixpkgs too. Fortunately Dune 2.5.0 has already been updated there.

Use `https://github.com/NixOS/nixpkgs/archive/807ca93fadd5197c2260490de0c76e500562dc05.tar.gz` as the URL and `10yq8bnls77fh3pk5chkkb1sv5lbdgyk1rr2v9xn71rr1k2x563p` as the sha256 in `dev/nixpkgs.nix`.

File: Makefile.dune (hunk 4 above)
Comment (ejgallego) [reply Makefile.dune:86]: Indeed, if that's your goal, `dune build @check` is more complete and much faster as it uses the non-optimizing compiler.

File: Makefile.dune (hunk 5 above)
Comment (ejgallego) [reply Makefile.dune:100]: Good catch; merge error.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
Do you know what is preventing the release on opam?

**Comment by ejgallego** [comment]
> Do you know what is preventing the release on opam?

Nothing special, some minor hiccup which is expected for a package so low in the dep graph.

**Comment by Zimmi48** [comment]
> > Do you know what is preventing the release on opam?
> 
> Nothing special, some minor hiccup which is expected for a package so low in the dep graph.

Not clear to me what you mean by this. I didn't see any PR open updating Dune on the opam-repository.

**Comment by ejgallego** [comment]
> Not clear to me what you mean by this. I didn't see any PR open updating Dune on the opam-repository.

https://github.com/ocaml/opam-repository/pull/16170#issuecomment-612207104

**Comment by ejgallego** [comment]
Upstream PR merged.

**Comment by ejgallego** [comment]
Unfortunately the Dune OCaml versioning policy is a bit too aggressive for us; that means:
- dune 2.5.0 can be installed in 4.05.0, however not at the same time than gtk due to conflicts
- a full install of dune requires 4.07.0

I've opted for only installing dune in 4.10.0; I dunno, we could move coqide testing to >= 4.07.0 and then install dune in 4.05.0 too, but neither solution is pretty and adding dune-configurator2.5.0 support for OCaml 4.05.0 is actually not a big deal but the way the library was written upstream makes this involved.

File: Makefile.dune (hunk 4 above)
Comment (ejgallego) [reply Makefile.dune:86]: My view is that these targets have been replaced by the shims, they do indeed depend on exactly what's needed for the desired use case.

**Comment by ejgallego** [comment]
Failure is spurious ; @rgrinberg just mentioned that they are planning to backport configurator 4.05.0 so that would be great. I think we just go ahead with this for now.

Actually I think that for 8.13 we should bump to 4.08 , but we have the performance bug and the linking bug which unfortunately are critical :/

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
Today is a great day :rainbow:

**Comment by proux01** [comment]
Great, we are now only waiting for ocaml/dune#3210 for this to be truly complete!

**Comment by SkySkimmer** [comment]
Looks like you broke azure (hidden by spurious failure here)?

**Comment by Zimmi48** [comment]
Damn, I'm sorry, I really should have restarted this one.

**Comment by SkySkimmer** [comment]
Also dune 2.5 isn't in the windows opam repo

**Comment by ejgallego** [comment]
As I said in the other PR `opam pin add dune --dev` may be a better fix until the opam repos is updated. This is what previous versions of this PR did to get green .

**Pull request merged**